### PR TITLE
nixos/binfmt: Add option to use static emulators when available

### PR DESCRIPTION
**Pull request by jcaesar**
## Motivation
This PR makes running binaries via `boot.binfmt.emulatedSystems` in chroots "just work", at the cost of adding/setting a single option.

The goal is to change
```
$ uname -m
x86_64
$ docker run --rm -ti --platform linux/arm alpine uname -m
exec /bin/uname: no such file or directory
```
into
```
$ docker run --rm -ti --platform linux/arm alpine uname -m
armv7l
```

The other prominent use-case is to `nixos-enter`/`chroot` into a broken linux install mounted to `/mnt` of an other machine for debugging, where that linux install is for a different platform (e.g. a raspi).

## Description of changes

* Add an option `boot.binfmt.preferStaticEmulators`, which changes `boot.binfmt.registrations.*.interpreter` to an executable from `pkgsStatic` where possible.
* Add a test for the option
* Some misc related cleanup

## Things I'd like advice on

Note how the error message in the motivational examples is abysmal? `/bin/uname` exists in both examples. What's not being found is `/run/binfmt/armv7l-linux`.

Instead of making this an option, would it be conceivable to make this always-on?

## Notes

This is the third attempt to land this in nixpkgs. Related PRs/issues:
* Original Issue: #160300
* First PR: #160802
  (abandoned)
* Second PR: #300070 
  (abandoned)
* Third PR: #314998
  (only made `pkgsStatic.qemu.override { userOnly = true; }` work)

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: lib/systems/default.nix [hunk 1]
```diff
@@ -261,23 +261,7 @@ let
...
-              pulseSupport = false;
-              pipewireSupport = false;
-              jackSupport = false;
-              smbdSupport = false;
-              seccompSupport = false;
-              tpmSupport = false;
-              capstoneSupport = false;
-              enableDocs = false;
+            qemu-user = pkgs.qemu-user.override {
               hostCpuTargets = [ "${final.qemuArch}-linux-user" ];
```
Comment (alyssais) [inline lib/systems/default.nix:265]: Do we want to keep this?  If we just used normal pkgs.qemu-user, we'd just need one Hydra-built binary for emulating any Linux platform.

Comment (jcaesar) [reply lib/systems/default.nix]: I thought I might get asked extra questions if I removed this. ;) Removed.

(Doing so does increase the necessary download/disk size for only enabling one/a handful of targets. If that's a concern, I could split outputs per architecture for qemu-user (and make the main `out`put a bunch of symlinks).)

**Comment by jcaesar** [comment]
I was hoping to get away without having to modify qemu again, but alas, `pkgsStatic.qemu-user` hasn't been merged for a week but already got broken. I added a commit that minifies dependencies so hopefully that will happen less quickly next time. The `qemu-user` and `qemu-utils` binaries (which had dependencies removed) are bit-for-bit equal before and after the change, so it shouldn't have broken anything.

---
Result of `nixpkgs-review pr 334859` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>5 packages marked as broken and skipped:</summary>
  <ul>
    <li>cot</li>
    <li>cot.dist</li>
    <li>ocamlPackages.ocaml-freestanding</li>
    <li>python312Packages.cot</li>
    <li>python312Packages.cot.dist</li>
  </ul>
</details>
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>8 packages failed to build:</summary>
  <ul>
    <li>lima-bin</li>
    <li>nemu</li>
    <li>qemu_full</li>
    <li>qemu_full.debug</li>
    <li>qemu_full.ga</li>
    <li>quickemu</li>
    <li>quickgui</li>
    <li>virtualboxKvm</li>
  </ul>
</details>
<details>
  <summary>86 packages built:</summary>
  <ul>
    <li>OVMFFull</li>
    <li>OVMFFull.fd</li>
    <li>alpine-make-vm-image</li>
    <li>cloud-init</li>
    <li>cloud-init.dist</li>
    <li>cloud-utils</li>
    <li>cloud-utils.guest</li>
    <li>colima</li>
    <li>diffoscope</li>
    <li>diffoscope.dist</li>
    <li>diffoscope.man</li>
    <li>gnome.gnome-boxes</li>
    <li>guestfs-tools</li>
    <li>libguestfs</li>
    <li>libguestfs-with-appliance</li>
    <li>lima</li>
    <li>linuxKernel.packages.linux_4_19.virtualbox</li>
    <li>linuxKernel.packages.linux_4_19_hardened.virtualbox</li>
    <li>linuxKernel.packages.linux_5_10.virtualbox</li>
    <li>linuxKernel.packages.linux_5_10_hardened.virtualbox</li>
    <li>linuxKernel.packages.linux_5_15.virtualbox</li>
    <li>linuxKernel.packages.linux_5_15_hardened.virtualbox</li>
    <li>linuxKernel.packages.linux_5_4.virtualbox</li>
    <li>linuxKernel.packages.linux_5_4_hardened.virtualbox</li>
    <li>linuxKernel.packages.linux_6_1.virtualbox</li>
    <li>linuxKernel.packages.linux_6_10.virtualbox</li>
    <li>linuxKernel.packages.linux_6_1_hardened.virtualbox</li>
    <li>linuxKernel.packages.linux_6_6.virtualbox</li>
    <li>linuxKernel.packages.linux_hardened.virtualbox (linuxKernel.packages.linux_6_6_hardened.virtualbox)</li>
    <li>linuxKernel.packages.linux_latest_libre.virtualbox</li>
    <li>linuxKernel.packages.linux_libre.virtualbox</li>
    <li>linuxKernel.packages.linux_lqx.virtualbox</li>
    <li>linuxKernel.packages.linux_xanmod.virtualbox</li>
    <li>linuxKernel.packages.linux_xanmod_latest.virtualbox (linuxKernel.packages.linux_xanmod_stable.virtualbox)</li>
    <li>linuxKernel.packages.linux_zen.virtualbox</li>
    <li>lxd-lts</li>
    <li>multipass</li>
    <li>nixpkgs-manual</li>
    <li>open-watcom-bin</li>
    <li>open-watcom-bin-unwrapped</li>
    <li>out-of-tree</li>
    <li>python311Packages.cot</li>
    <li>python311Packages.cot.dist</li>
    <li>python311Packages.guestfs</li>
    <li>python311Packages.guestfs.dist</li>
    <li>python312Packages.guestfs</li>
    <li>python312Packages.guestfs.dist</li>
    <li>qemu</li>
    <li>qemu-user</li>
    <li>qemu-user.debug</li>
    <li>qemu-utils</li>
    <li>qemu-utils.debug</li>
    <li>qemu.debug</li>
    <li>qemu.ga</li>
    <li>qemu_kvm</li>
    <li>qemu_kvm.debug</li>
    <li>qemu_kvm.ga</li>
    <li>qemu_test</li>
    <li>qemu_test.debug</li>
    <li>qemu_test.ga</li>
    <li>qemu_xen (qemu_xen_4_19)</li>
    <li>qemu_xen.debug (qemu_xen_4_19.debug)</li>
    <li>qemu_xen.ga (qemu_xen_4_19.ga)</li>
    <li>qemu_xen_4_16</li>
    <li>qemu_xen_4_16.debug</li>
    <li>qemu_xen_4_16.ga</li>
    <li>qemu_xen_4_17</li>
    <li>qemu_xen_4_17.debug</li>
    <li>qemu_xen_4_17.ga</li>
    <li>qemu_xen_4_18</li>
    <li>qemu_xen_4_18.debug</li>
    <li>qemu_xen_4_18.ga</li>
    <li>qtemu</li>
    <li>solo5</li>
    <li>solo5.debug</li>
    <li>vagrant</li>
    <li>virtualbox</li>
    <li>virtualbox.modsrc</li>
    <li>virtualboxHardened</li>
    <li>virtualboxHardened.modsrc</li>
    <li>virtualboxHeadless</li>
    <li>virtualboxHeadless.modsrc</li>
    <li>virtualboxWithExtpack</li>
    <li>virtualboxWithExtpack.modsrc</li>
    <li>vmctl</li>
    <li>zpool-auto-expand-partitions</li>
  </ul>
</details>

File: nixos/modules/system/boot/binfmt.nix [hunk 2]
```diff
@@ -280,28 +278,50 @@ in {
...
     boot.binfmt.registrations = builtins.listToAttrs (map (system: assert system != pkgs.stdenv.hostPlatform.system; {
       name = system;
       value = { config, ... }: let
-        interpreter = getEmulator system;
-        qemuArch = getQemuArch system;
+        elaborated = lib.systems.elaborate { inherit system; };
+        useStaticEmulator = cfg.preferStaticEmulators && elaborated.staticEmulatorAvailable pkgs;
+        interpreter = elaborated.emulator (if useStaticEmulator then pkgs.pkgsStatic else pkgs);
+
+        qemuArch = elaborated.qemuArch;
```
Comment (SuperSandro2000) [inline nixos/modules/system/boot/binfmt.nix:310]:
```suggestion
        inherit (elaborated) qemuArch;
```

**Review by SuperSandro2000 — APPROVED** [review]

File: pkgs/applications/virtualization/qemu/default.nix [hunk 3]
```diff
@@ -86,10 +86,9 @@ stdenv.mkDerivation (finalAttrs: {
     # workaround, remove once this patch lands: https://lore.kernel.org/qemu-devel/20240805104921.4035256-1-hi@alyssa.is/
     ++ lib.optionals (hexagonSupport && stdenv.hostPlatform.isStatic) [ pcre2 ];
 
-  buildInputs = [ zlib glib pixman
-    vde2 lzo snappy libtasn1
-    gnutls nettle curl libslirp
-  ]
+  buildInputs = [ glib zlib ]
+    ++ lib.optionals (!minimal) [ dtc pixman vde2 lzo snappy libtasn1 gnutls nettle libslirp ]
+    ++ lib.optionals (!userOnly) [ curl ]
```
Comment (SuperSandro2000) [inline pkgs/applications/virtualization/qemu/default.nix:91]: I am not really sure how to verify this but disabling features we don't need is probably a good idea.

Comment (jcaesar) [reply pkgs/applications/virtualization/qemu/default.nix:91]: What I did:
1. Verify manually that for `!minimal && !userOnly` (i.e. all the normal qemus like qemu, qemu-full, qemu_kvm…), only the order of inputs changes (which is hopefully irrelevant).
2. qemu-user working as intended is covered by a test added here
3. For qemu-utils:
   * `nix build github:NixOS/nixpkgs/{c46d43c625f161d3bd22052a5f8f9df703d81be8,f69411bcfb0dd3565a19667fdf509bbadb38f008}#qemu-utils`
   * The sizes of the output binaries are all the same: `ls -l result*/bin/* | cut -d\  -f5- | sort`, so it's unlikely that some larger functionality went missing
   * Looking e.g. through `nix run nixpkgs#diffoscope result*/bin/qemu-storage-daemon` shows some probably irrelevant differences in the length of some dummy strings (Probably due to `error: derivation '…-qemu-utils-9.0.2.drv' may not be deterministic: output '…-qemu-utils-9.0.2' differs` -.-)

**Review by SuperSandro2000 — APPROVED** [review]

**Review by Cryolitia — APPROVED** [review]

**Comment by alyssais** [comment]
lib and pkgs changes look fine to me.

File: nixos/modules/system/boot/binfmt.nix [hunk 4]
```diff
@@ -280,28 +278,50 @@ in {
         '';
         type = types.listOf (types.enum (builtins.attrNames magics));
       };
+
+      preferStaticEmulators = mkOption {
+        default = "false";
```
Comment (SuperSandro2000) [inline nixos/modules/system/boot/binfmt.nix:283]:
```suggestion
        default = false;
```
Shouldn't this be a bool?

**Review by SuperSandro2000 — APPROVED** [review]
otherwise LGTM and I would say we merge this

**Comment by jcaesar** [comment]
@SuperSandro2000 Can I do something to make this easier to merge? Maybe split off the changes to pkgs/ into a different PR? (They're not strictly necessary, but should help to avoid test build failures.) Or is this simply ready to hit the button?

---

Rebased due to merge conflicts. nemu and quickgui fail as before.

---

Result of `nixpkgs-review` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>8 packages marked as broken and skipped:</summary>
  <ul>
    <li>cot</li>
    <li>cot.dist</li>
    <li>linuxKernel.packages.linux_5_4_hardened.virtualbox</li>
    <li>ocamlPackages.ocaml-freestanding</li>
    <li>python311Packages.subunit2sql</li>
    <li>python311Packages.subunit2sql.dist</li>
    <li>python312Packages.cot</li>
    <li>python312Packages.cot.dist</li>
  </ul>
</details>
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>4 packages failed to build:</summary>
  <ul>
    <li>nemu</li>
    <li>quickgui</li>
    <li>quickgui.debug</li>
    <li>quickgui.pubcache</li>
  </ul>
</details>
<details>
  <summary>157 packages built:</summary>
  <ul>
    <li>OVMFFull</li>
    <li>OVMFFull.fd</li>
    <li>alpine-make-vm-image</li>
    <li>barbicanclient (python311Packages.python-barbicanclient)</li>
    <li>barbicanclient.dist (python311Packages.python-barbicanclient.dist)</li>
    <li>cloud-init</li>
    <li>cloud-init.dist</li>
    <li>cloud-utils</li>
    <li>cloud-utils.guest</li>
    <li>colima</li>
    <li>diffoscope</li>
    <li>diffoscope.dist</li>
    <li>diffoscope.man</li>
    <li>glanceclient (python311Packages.python-glanceclient)</li>
    <li>glanceclient.dist (python311Packages.python-glanceclient.dist)</li>
    <li>gnome-boxes</li>
    <li>guestfs-tools</li>
    <li>heatclient (python311Packages.python-heatclient)</li>
    <li>heatclient.dist (python311Packages.python-heatclient.dist)</li>
    <li>ironicclient (python311Packages.python-ironicclient)</li>
    <li>ironicclient.dist (python311Packages.python-ironicclient.dist)</li>
    <li>kata-runtime</li>
    <li>libguestfs</li>
    <li>libguestfs-with-appliance</li>
    <li>lima</li>
    <li>lima-bin</li>
    <li>linuxKernel.packages.linux_5_10.virtualbox</li>
    <li>linuxKernel.packages.linux_5_10_hardened.virtualbox</li>
    <li>linuxKernel.packages.linux_5_15.virtualbox</li>
    <li>linuxKernel.packages.linux_5_15_hardened.virtualbox</li>
    <li>linuxKernel.packages.linux_5_4.virtualbox</li>
    <li>linuxKernel.packages.linux_6_1.virtualbox</li>
    <li>linuxKernel.packages.linux_6_10.virtualbox</li>
    <li>linuxKernel.packages.linux_6_11.virtualbox</li>
    <li>linuxKernel.packages.linux_6_1_hardened.virtualbox</li>
    <li>linuxKernel.packages.linux_6_6.virtualbox</li>
    <li>linuxKernel.packages.linux_hardened.virtualbox (linuxKernel.packages.linux_6_6_hardened.virtualbox)</li>
    <li>linuxKernel.packages.linux_latest_libre.virtualbox</li>
    <li>linuxKernel.packages.linux_libre.virtualbox</li>
    <li>linuxKernel.packages.linux_lqx.virtualbox</li>
    <li>linuxKernel.packages.linux_xanmod.virtualbox</li>
    <li>linuxKernel.packages.linux_xanmod_latest.virtualbox (linuxKernel.packages.linux_xanmod_stable.virtualbox)</li>
    <li>linuxKernel.packages.linux_zen.virtualbox</li>
    <li>lxd-lts</li>
    <li>magnumclient (python311Packages.python-magnumclient)</li>
    <li>magnumclient.dist (python311Packages.python-magnumclient.dist)</li>
    <li>manilaclient (python311Packages.python-manilaclient)</li>
    <li>manilaclient.dist (python311Packages.python-manilaclient.dist)</li>
    <li>mistralclient (python311Packages.python-mistralclient)</li>
    <li>mistralclient.dist (python311Packages.python-mistralclient.dist)</li>
    <li>multipass</li>
    <li>nixpkgs-manual</li>
    <li>open-watcom-bin</li>
    <li>open-watcom-bin-unwrapped</li>
    <li>openstackclient (python311Packages.python-openstackclient)</li>
    <li>openstackclient-full</li>
    <li>openstackclient-full.dist</li>
    <li>openstackclient.dist (python311Packages.python-openstackclient.dist)</li>
    <li>out-of-tree</li>
    <li>python311Packages.cot</li>
    <li>python311Packages.cot.dist</li>
    <li>python311Packages.guestfs</li>
    <li>python311Packages.guestfs.dist</li>
    <li>python311Packages.keystoneauth1</li>
    <li>python311Packages.keystoneauth1.dist</li>
    <li>python311Packages.openstacksdk</li>
    <li>python311Packages.openstacksdk.dist</li>
    <li>python311Packages.openstacksdk.man</li>
    <li>python311Packages.os-client-config</li>
    <li>python311Packages.os-client-config.dist</li>
    <li>python311Packages.osc-lib</li>
    <li>python311Packages.osc-lib.dist</li>
    <li>python311Packages.osc-placement</li>
    <li>python311Packages.osc-placement.dist</li>
    <li>python311Packages.oslo-concurrency</li>
    <li>python311Packages.oslo-concurrency.dist</li>
    <li>python311Packages.oslo-db</li>
    <li>python311Packages.oslo-db.dist</li>
    <li>python311Packages.oslo-log</li>
    <li>python311Packages.oslo-log.dist</li>
    <li>python311Packages.oslo-serialization</li>
    <li>python311Packages.oslo-serialization.dist</li>
    <li>python311Packages.oslo-utils</li>
    <li>python311Packages.oslo-utils.dist</li>
    <li>python311Packages.osprofiler</li>
    <li>python311Packages.osprofiler.dist</li>
    <li>python311Packages.python-aodhclient</li>
    <li>python311Packages.python-aodhclient.dist</li>
    <li>python311Packages.python-cinderclient</li>
    <li>python311Packages.python-cinderclient.dist</li>
    <li>python311Packages.python-designateclient</li>
    <li>python311Packages.python-designateclient.dist</li>
    <li>python311Packages.python-keystoneclient</li>
    <li>python311Packages.python-keystoneclient.dist</li>
    <li>python311Packages.python-neutronclient</li>
    <li>python311Packages.python-neutronclient.dist</li>
    <li>python311Packages.python-novaclient</li>
    <li>python311Packages.python-novaclient.dist</li>
    <li>python311Packages.python-octaviaclient</li>
    <li>python311Packages.python-octaviaclient.dist</li>
    <li>swiftclient (python311Packages.python-swiftclient)</li>
    <li>swiftclient.dist (python311Packages.python-swiftclient.dist)</li>
    <li>troveclient (python311Packages.python-troveclient)</li>
    <li>troveclient.dist (python311Packages.python-troveclient.dist)</li>
    <li>watcherclient (python311Packages.python-watcherclient)</li>
    <li>watcherclient.dist (python311Packages.python-watcherclient.dist)</li>
    <li>python311Packages.python-zaqarclient</li>
    <li>python311Packages.python-zaqarclient.dist</li>
    <li>zunclient (python311Packages.python-zunclient)</li>
    <li>zunclient.dist (python311Packages.python-zunclient.dist)</li>
    <li>python311Packages.swift</li>
    <li>python311Packages.swift.dist</li>
    <li>python311Packages.tempest</li>
    <li>python311Packages.tempest.dist</li>
    <li>python312Packages.guestfs</li>
    <li>python312Packages.guestfs.dist</li>
    <li>qemu</li>
    <li>qemu-user</li>
    <li>qemu-user.debug</li>
    <li>qemu-utils</li>
    <li>qemu-utils.debug</li>
    <li>qemu.debug</li>
    <li>qemu.ga</li>
    <li>qemu_full</li>
    <li>qemu_full.debug</li>
    <li>qemu_full.ga</li>
    <li>qemu_kvm</li>
    <li>qemu_kvm.debug</li>
    <li>qemu_kvm.ga</li>
    <li>qemu_test</li>
    <li>qemu_test.debug</li>
    <li>qemu_test.ga</li>
    <li>qemu_xen</li>
    <li>qemu_xen.debug</li>
    <li>qemu_xen.ga</li>
    <li>qemu_xen_4_17</li>
    <li>qemu_xen_4_17.debug</li>
    <li>qemu_xen_4_17.ga</li>
    <li>qemu_xen_4_18</li>
    <li>qemu_xen_4_18.debug</li>
    <li>qemu_xen_4_18.ga</li>
    <li>qtemu</li>
    <li>quickemu</li>
    <li>solo5</li>
    <li>solo5.debug</li>
    <li>vagrant</li>
    <li>virtualbox</li>
    <li>virtualbox.modsrc</li>
    <li>virtualboxHardened</li>
    <li>virtualboxHardened.modsrc</li>
    <li>virtualboxHeadless</li>
    <li>virtualboxHeadless.modsrc</li>
    <li>virtualboxKvm</li>
    <li>virtualboxWithExtpack</li>
    <li>virtualboxWithExtpack.modsrc</li>
    <li>vmctl</li>
    <li>zpool-auto-expand-partitions</li>
  </ul>
</details>

**Comment by SuperSandro2000** [comment]
Looks all good to me. Just give me a ping when you fixed the test and as long as the changes done there are looking good, I would just merge this.

**Comment by jcaesar** [comment]
The test failure on linux seems to be flaky, I couldn't reproduce it in 5 tries, a no-change force push made it go away. (darwin times out as always.)

**Review by SuperSandro2000 — APPROVED** [review]

**Comment by jcaesar** [comment]
@oryjkov
I don't have an `aarch64` system, but people were able to build this before (and then it failed elsewhere), see [here](https://github.com/NixOS/nixpkgs/issues/366902).
In any case, it looks like you want to open a separate issue?

**Pull request merged**